### DISCUSSION TITLE
compare more-info manipulation in custom-ui

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [NemesisRE/kiosk-mode#98] [Ideas]
HI,

please let me suggest some form of mutual checks ;-)

in c[ustom-ui](https://github.com/Mariusthvdb/custom-ui/tree/master), we ve been handling the more-info options for a very long time. More specifically, by means of a [`hide_attributes` ](https://github.com/Mariusthvdb/custom-ui/blob/master/EXAMPLES.md#hide-attributes) option in customize, and have that change the contents of the more-info attributes list in the dropdown.

This was introduced to not only hide specific attributes, but also to hide the template that custom-ui adds to entities.

It has been a challenge to have that code update with the HA Frontend changes, as you are aware....

Since I am not a dev at all, and merely try to keep that JS script alive, I was hoping you could have a look at it, and see if some code could be change, to make it behave better. The dom paths and traversing those to the more-info panels, and group more info panels are of special interest in my custom-ui.

Currently it works nicely. with 1 big caveat and 1 smaller:

the smaller:
one can see the script in action: upon clicking the dropdown, one can see the box be made smaller and hiding the attributes, from an initial bigger dropdown.

![Jun-28-2023 10-34-55](https://github.com/NemesisRE/kiosk-mode/assets/33354141/85b89af9-bed4-40d2-af3b-c06d3c7d52ca)

the bigger:
if, because of the hide_attributes, no attributes are left to show, the dropdown is still displayed in the more-info, albeit with an empty box. Ive tried to prevent that, by adding some logic to the core attributes filter, but havent been able to manage that.

Ive even had some brief chat with the core frontend devs on this, but I cant make it happen see https://github.com/Mariusthvdb/custom-ui/discussions/52

I would appreciate you having a look enormously, and would love to get any feedback you might have.

__update__

I've separated the function into a dedicated resource here: https://github.com/Mariusthvdb/custom-ui/blob/master/update_more_info.js

**Comment by elchininet** [comment]
Hi @Mariusthvdb,
I need to take a look closer at the code because I am not releated to it at all. I have not used that plugin before so first I need to understand how it works so I can give you feedback about it and if it can be improved in some way.
Regards

**Reply by Mariusthvdb** [reply]
Ofc, take your time, no hurry at all.

I truly appreciate anything you could spot, or time you would spend.

**Reply by elchininet** [reply]
Hi @Mariusthvdb,

I tried to read the code on the repo and I am seeing that [there is a logic](https://github.com/Mariusthvdb/custom-ui/blob/0b0804ede8cb6f212722ab83d7f5810440e9f190/custom-ui.js#L116) to add `display none` to attributes, but that logic [runs when a certain event is triggered](https://github.com/Mariusthvdb/custom-ui/blob/0b0804ede8cb6f212722ab83d7f5810440e9f190/custom-ui.js#L193). It is logic that you see that small jump, because the items are rendered and then the logic runs trying to hide them. Maybe it could be overcomed if you know in advance what will be rendered inside the attributes, but I don‘t know if that could be possible.

The second I cannot get it just reading the code. You can do something, try to copy/paste the [Docker demo commands from this repo](https://github.com/NemesisRE/kiosk-mode/blob/bb1e5b4ba203c87c77fef596d7d1fb0e3dce1958/package.json#L16-L20) (adapting them to `custom-ui`) and [a config](https://github.com/NemesisRE/kiosk-mode/tree/master/.hass/config) that sets up some examples of the usage of `custom-ui`. In that way it is very easy to run the demo command and get a Home Assistant instance with some `custom-ui` features already configured and the issues could be reproduced there, in that way I can check what is happening in live instead of trying to get it just reading the code. I can give you more feedback if I can reproduce the issues in a fresh HA instance.

Regards

**Comment by Mariusthvdb** [comment]
right, now 2023.9 beta has landed my more-info update custom code no longer works.
Not sure if they changed the event (probably not) or the dom paths, but dropping down the attributes box does not change anything anymore in my card.

ive separated the code into a dedicated custom card for debugging purposes:
```
function updateMoreInfo(ev) {
  if (!ev.detail.expanded) return;
  const moreInfoInfo = document
    .querySelector("home-assistant")
    .shadowRoot.querySelector("ha-more-info-dialog")
    .shadowRoot.querySelector("ha-dialog")
    .getElementsByClassName("content")[0]
    .querySelector("ha-more-info-info");

  try {
    let t;
    {
      let moreInfoNodeName;
      const contentChild = moreInfoInfo.shadowRoot.querySelector(
        "more-info-content"
      ).childNodes;
      for (const nodeItem of contentChild) {
        if (nodeItem.nodeName.toLowerCase().startsWith("more-info-")) {
          moreInfoNodeName = nodeItem.nodeName.toLowerCase();
          break;
        }
      }
      if (moreInfoNodeName == "more-info-group") {
        let moreInfoNestedNodeName;
        const contentChildNested =
          moreInfoInfo.shadowRoot
            .querySelector("more-info-group")
            .shadowRoot.childNodes;
        for (const nodeItemNested of contentChildNested) {
          if (
            nodeItemNested.nodeName.toLowerCase().startsWith("more-info-")
          ) {
            moreInfoNestedNodeName = nodeItemNested.nodeName.toLowerCase();
            break;
          }
        }
        t = moreInfoInfo.shadowRoot
          .querySelector("more-info-group")
          .shadowRoot.querySelector(moreInfoNestedNodeName)
          .shadowRoot.querySelector("ha-attributes")
          .shadowRoot.querySelectorAll(".data-entry");
      } else {
        t = moreInfoInfo.shadowRoot
          .querySelector(moreInfoNodeName)
          .shadowRoot.querySelector("ha-attributes")
          .shadowRoot.querySelectorAll(".data-entry");
      }
    }
    if (t.length) {
      let e;
      for (const node of t) {
        const o = node.getElementsByClassName("key")[0];
        if (o.innerText.toLowerCase().trim() == "hide attributes") {
          e = o.parentNode
            .getElementsByClassName("value")[0]
            .innerText.split(",")
            .map((item) => item.replace("_", " ").trim());
          e.push("hide attributes");
        }
      }
      for (const node of t) {
        const o = node.getElementsByClassName("key")[0];
        if (
          e.includes(o.innerText.toLowerCase().trim()) ||
          e.includes("all")
        ) {
          o.parentNode.style.display = "none";
        }
      }
    }
  } catch (err) {}
}

window.addEventListener("expanded-changed", updateMoreInfo);
```

which works as expected in 2023.8.4

I even had hopes changing that constant (dom path) in MoreInfoInfo to 
```
function updateMoreInfo(ev) {
  if (!ev.detail.expanded) return;
  const expansionPanel = document.querySelector("ha-expansion-panel");
```

and start from there, but that doesnt work in 2023.8 either. Would have been nicer, and more robust.

Im out of my league here, so if you could see what's wrong in this code reaching attributes in the more-info panels on regular entities and groups (because was the challenge before) I'd still be very appreciative ...

Didnt test whether Kiosk mode worked ok in 2023.9 beta, because I had to downgrade seeing system blocking issues.

**Reply by elchininet** [reply]
Hi @Mariusthvdb,
I‘ll try to see if I can help you in some way. What should I expect from that code? How do I know if it is working as expected?
Regards

**Comment by Mariusthvdb** [comment]
wow thanks!

short: set a customization `hide_attributes` under customize on any entity, and make it disappear from the attributes list in more-info.

this was originally introduced in custom-ui, to hide the attribute `templates` that allows on customizations. It as expanded a bit to hide itself also, hide any other attribute, or hide all ;-)

you can see what is does at https://github.com/Mariusthvdb/custom-ui/blob/master/EXAMPLES.md#hide-attributes

```
  customize_glob:
    # All Entities - Hide Templates & Icon Color
    # Note that yaml keys can't start with an asterisk. Use quotes in that case:
    "*.*":
      hide_attributes:
        - templates
        - icon_color
        - all # <-- this hides, well, all attributes ;-)
```

or like I do on all individual entities that get a template:

```
homeassistant:

  customize:

    binary_sensor.espresso_ready:
      hide_attributes: templates
      templates:
        icon_color: >
          return state === 'on' ? 'rgb(93,53,15)' : 'var(--primary-color)';
```

(this shows the famous icon_color option, which is the other main feature of custom-ui)

hope this is clear, if not, please let me know.

**Reply by elchininet** [reply]
Ok, let, me install it first and I will check in which part it is failing.
I‘ll let you know.
Regards

**Reply by elchininet** [reply]
When should the `expanded-changed` event be triggered?
It is not triggered in my HA instance `2023.8.4`.

**Reply by elchininet** [reply]
I got it, the event is triggered when the attributes are opened.

**Reply by Mariusthvdb** [reply]
tbh, my long-standing wish is to add the attributes set under the hide_attributes key in customize be somehow added to the filter list HA itself uses to hide/show attributes. Ive had coms with Bram on that some time ago, but never understood how to make that work.

What happens now is, that when the hide_attributes filters a final attribute, the dropdown box is still shown, albeit empty.... while in core, when no attributes are left to show, the box itself is not shown.

that would probably be something you'd understand the inner workings of, its beyond my capabilities in all honesty.

let me find what Bram said about that

this:
```
const jouwFilters = "attribute1, attribute2, attribute3"
// de queryselector moet je zelf ff zoeken, maar die zou je al moeten hebben
const haAttributes = querySelector("ha-attributes");
haAttributes.extraFilters = haAttributes.extraFilters + (haAttributes.extraFilters.length ? "," : "") + jouwFilters;```

**Reply by Mariusthvdb** [reply]
long time ago we would find that list in https://github.com/home-assistant/frontend/blob/7bc27082595f943d15ca35ae948c766184a93eed/src/util/hass-attributes-util.ts#L87

but I dont think the file exists in modern Frontend anymore.

finally found at least the source referencing those, in https://github.com/home-assistant/frontend/blob/dev/src/components/ha-attributes.ts

**Reply by elchininet** [reply]
@Mariusthvdb,
Try to change this:

```javascript
e = o.parentNode
    .getElementsByClassName("value")[0]
    .innerText.split(",")
    .map((item) => item.replace("_", " ").trim());
e.push("hide attributes");
```

By this:

```javascript
e = o.parentNode
    .getElementsByClassName("value")[0]
    .querySelector('ha-attribute-value')
    .shadowRoot
    .textContent.split(",")
    .map((item) => item.replace("_", " ").trim());
e.push("hide attributes");
```

**Reply by Mariusthvdb** [reply]
thanks! before I do, did you try this on 2023.8 or the beta 2023.9

just asking because I downgraded and wouldn't be able to test if this was a 2023.9 fix only. Seems to not work on 2023.8.4, not hiding the templates, or the hide_attributes line itself.

btw, [Karwosts mentioned](https://discord.com/channels/330944238910963714/351047592588869643/1146913635407515648) in #devs_frontend, I should probably be looking to change the `STATE_ATTRIBUTES ` in https://github.com/home-assistant/frontend/blob/6c7efc17c2b7dd3306b3577f615b267fd274f2ba/src/data/entity_attributes.ts#L1 and mentioned in https://github.com/home-assistant/frontend/blob/6c7efc17c2b7dd3306b3577f615b267fd274f2ba/src/components/ha-attributes.ts#L30

maybe you might be able to see how we could simply apply Brams suggestion above to that `extraFilters` in https://github.com/home-assistant/frontend/blob/dev/src/components/ha-attributes.ts#L28? as Karwosts [mentions](https://discord.com/channels/330944238910963714/351047592588869643/1146915150671134771)

thanks Iván, much appreciated!

**Reply by Mariusthvdb** [reply]
Update: I took the plunge and updated to 2023.9b01, and can confirm the edit to work !

Hiding the `templates` attribute and the `hide_attributes` line itself nicely. that is really great!!

if you have some time I really would appreciate to learn how you found that.

**Reply by elchininet** [reply]
Hi @Mariusthvdb,

If it works, then I will give you a code that works for both versions with a commented block that you can delete when all you users have migrated. I'll prepare it later.

To detect the issue I just created a new instance of HA using the test Docker image from `kiosk-mode` and I started to read the code and check the DOM elements until I noticed a breaking change in the new version, so your current code fails silently under the beta.

Regards

**Reply by Mariusthvdb** [reply]
magic.
if I may ask something please:
if I understand correctly, the function `UpdateMoreInfo` currently is called each and every time the attributes dropdown box is `expanded-changed`.
it then hide the attributes, if it finds the `hide_attributes:` attribute and does its magic.

Isnt this causing unnecessary action on entities that _dont_ have that `hide_attributes` attribute set. I mean, couldnt this logic be changed, to be only updated if and when it actually has that attribute? Ignore otherwise.

If possible wouldn't it be better to get that if in front of the whole functionality, even before sensing the `expanded-changed`

**Reply by elchininet** [reply]
>if I understand correctly, the function UpdateMoreInfo currently is called each and every time the attributes dropdown box is expanded-changed.

You got it right, it is in that way

>Isnt this causing unnecessary action on entities that dont have that hide_attributes attribute set. I mean, couldnt this logic be changed, to be only updated if and when it actually has that attribute? Ignore otherwise.

I dont‘t know Home Assistant front-end code at a level that I could tell you if that would be possible or not, but most probably it was done in that way because it was impossible to retrieve that information in advance.

**Reply by Mariusthvdb** [reply]
this might be vital, I just had to downgrade again because 2023.9 (now on b01) keeps crashing Zwave, and we can not miss out on that.

If at all possible, would you be able to implement the code so, that it auto-checks the HA version and selects the right code for it? We [had that before in our code at sometime in the past](https://github.com/Mariusthvdb/custom-ui/releases/tag/20210322-X), but it was taken out. Don't want to bug users with something trivial like that....

This seems an important moment to reintroduce that backwards compatibility while going forward

**Reply by elchininet** [reply]
Hi @Mariusthvdb,

Here you go, this should work in both versions. But you need to be aware that this "retro-compatibility" will last until they decide to change the DOM structure again 😉

```javascript
if (o.innerText.toLowerCase().trim() == "hide attributes") {
    const valueContainer = o.parentNode.getElementsByClassName("value")[0];
    const haAttributeValue = valueContainer.querySelector('ha-attribute-value');
    const text = haAttributeValue
        ? haAttributeValue.shadowRoot.textContent
        : valueContainer.innerText;
    e = text
        .split(",")
        .map((item) => item.replace("_", " ").trim());
    e.push("hide attributes");
}
```

**Reply by Mariusthvdb** [reply]
tested and confirmed to work in 2023.8
hope to update to 2023.9 tomorrow, so will provide feedback,

thanks!

if you'd care to PR the resources, Id be happy to merge them in both the custom-ui.js, and the separate custom-more-info.js and acknowledge your work

**Reply by Mariusthvdb** [reply]
Hi Ivan @elchininet 
I fear its that time again...

hide templates in more-info is no longer working.

besides that, all of a sudden, the customized templates we set in custom-ui on attributes are no longer template able in jinja. see: https://github.com/home-assistant/core/issues/104190

Seems there are some changes in HA core I can not mitigate myself. I would really appreciate it if you could have another check why the more info doesnt work anymore, the path seems still valid??


here's the issue in custom-ui repo https://github.com/Mariusthvdb/custom-ui/issues/111

**Reply by elchininet** [reply]
Hi @Mariusthvdb,
Sure, I‘ll take a look at it to see if I can spot where is the issue.
Meanwhile take a look at [this](https://github.com/elchininet/home-assistant-query-selector). I am thinking on mitigating all of these Home Assistant changes from a single place.
Regards

**Reply by Mariusthvdb** [reply]
o wow that looks very promising, need to inspect that further, but I see a lot of related conceps there ;-)
will follow with utmost attention.


btw the more-info seems to be not broken after all, I suspect it to be a cache issue of my testing the other open issue in my repo https://github.com/Mariusthvdb/custom-ui/issues/110

maybe that too is a Path problem, so if you'd be able to coach that instead...... ;-)

**Reply by Mariusthvdb** [reply]
btw `Lovelave` probably is a typo in your code?

**Reply by elchininet** [reply]
> btw Lovelave probably is a typo in your code?

Yes, it is, good catch 😉

**Reply by elchininet** [reply]
>maybe that too is a Path problem, so if you'd be able to coach that instead...... ;-)

Could you sumarising what is the issue there? I got lost in the discussion :)

**Reply by Mariusthvdb** [reply]
The issue is that when we set a template in the customize section, that template is correctly added to the attributes but the result . Eg icon_color is no longer recognized as such. 
Before , we could do {{state_attr('sensor.xyz','icon_color')}} and it would return the outcome of that template in customize. Now it returns null . Says type is dict.

The icon_color Is displayed correctly though in the icon, so we're 100% positive that still works.

It must be something changed in the attributes handling and subsequently templates can't evaluate.

See the linked issue in my repo

**Reply by elchininet** [reply]
@Mariusthvdb,

Remember that I don‘t have any clue of how `custom-ui` works because I don‘t use it myself, I see the screenshots in the issue and I don‘t see anything wrong because I don‘t know what is expected. I see that the icon gets coloured and in the attributes section the colour of the icon appears correctly.

As before, just send me the piece of code that is failing, and let me know what do you expect and what are you getting instead. I can help you better if I can reproduce it in my local.

**Reply by Mariusthvdb** [reply]
thats is just the thing, I dont now which code in the custom-ui now is failing.... as I take it its a HA code change.

main thing is that even though

> in the attributes section the colour of the icon appears correctly.

the attribute can no longer be templated in a jinja templated, resulting in Null.

before, jinja templates could correctly see te outcome of any template we set in custom-ui, and that has gone now.
(custom-ui can set templates on any attribute (existing or additional), not only icon_color)

**Reply by elchininet** [reply]
Hi @Mariusthvdb,
I was trying to understand a bit the logic in `custom-ui` to see if I can help you out with your issue. Have you played around with [this block](https://github.com/Mariusthvdb/custom-ui/blob/95293301f9a316eca718b575a1808914a7a0c865/custom-ui.js#L214-L228) and [this block of code](https://github.com/Mariusthvdb/custom-ui/blob/95293301f9a316eca718b575a1808914a7a0c865/custom-ui.js#L15-L47)?
If something is failing in the template rendering logic it should be there.
Regards

**Reply by Mariusthvdb** [reply]
I havent yet no. I didn't not find an issue yet...
Thanks for looking nine to this, I really appreciate that!!


at first I believed the last HA version to break something, but, as I went back and downgraded 2 versions, the issue remained....

I also downgraded custom-ui to the original and several later updates, but still the issue remains.

Maybe it has been like this always, and we just have bad memory....

Anyways, I did issue a new version, without templating, and only adding icon_color (to be used in customize, And in the attributes section of regular template entities.)

The last version of regular custom-ui does works otherwise, and is super powerful, so I guess we should leave it at that, unless someone with a true dev eye (not me, I merely kept it alive...) has a fix for it.

If you want to have a glimpse at the new version, please do, it is extremely small ;-)
https://github.com/Mariusthvdb/custom-ui/releases/tag/20231123

**Reply by elchininet** [reply]
Glad that it is working.

In the reduced version just add `window.customUI.initDone = true;` in some place, it seems that you want to prevent initialising everything once it was already done with `if (window.customUI.initDone) return;` but you are not setting that variable to `true` later, so it will pass that condition all the time.

**Reply by Mariusthvdb** [reply]
thanks, I admit not having checked that, and only taken part the original....

maybe, if you d be willing, continue any discussion on my repo? that way we wont liter yours with specifics not having to do anything with Kiosk mode....

**Reply by Mariusthvdb** [reply]
here's a new repo, and Ive mentioned you in the first update ;-)
https://github.com/Mariusthvdb/custom-ui-icon-color/tree/main

fyi, Ive also added a card-mod-examples page, geared specifically to those custom-ui mods for icon /  icon_color and how to use these in various cards/combinations

**Reply by elchininet** [reply]
>maybe, if you d be willing, continue any discussion on my repo? that way we wont liter yours with specifics not having to do anything with Kiosk mode....

Yes, just ping me in any issue or discussion and I‘ll glad to continue there.

**Comment by Mariusthvdb** [comment]
can finally confirm succes in the 2023.9b05 update.

all more-info's are nicely updated, and attributes are hidden

will update resources hopefully tomorrow!

**Reply by elchininet** [reply]
Good to know @Mariusthvdb 👍
I saw that you were searching in Discord a way to access the config of entities, did you find something?

**Reply by Mariusthvdb** [reply]
No, unfortunately I havent, and is seems to utterly simple, especially now these extraFilters have moved, and should probably be addressable as such, (without having to deal with a Dom path at all?)

My current (fixed by you) script is so clunky, reminiscing the old days. It hurts my eyes, but there is no other way for now.

I really think that is the only way forward, simply add those attributes to that list, and have the HA core  frontend take care of them. 
it's just me not being capable enough to find out how..
see: https://github.com/Mariusthvdb/custom-ui/discussions/99#discussioncomment-6886350

**Comment by Mariusthvdb** [comment]
might I ping you on an issue I now found:
```
Logger: frontend.js.latest.202309080
Source: components/system_log/__init__.py:269 
First occurred: 10 september 2023 om 16:54:29 (30 occurrences) 
Last logged: 07:48:52

Uncaught error from Safari 16.6 on Mac OS 10.15.7 Script error. null @:0:0
Uncaught error from Safari 16.6 on Mac OS 10.15.7 TypeError: null is not an object (near '...elector("ha-dialog")...') updateMoreInfo (/local/resources/custom-ui/custom-more-info.js:19:43) dispatchEvent ([native code]::) event (src/common/dom/fire_event.ts:76:21) nextRender (src/components/ha-expansion-panel.ts:131:14)
```

didnt spot this when testing, and havent yet found a reason/action for it to occur but maybe you understand point blank what this means?
I cant drop the .js file here, but it is in my repo at https://github.com/Mariusthvdb/custom-ui/blob/master/update_more_info.js

opened an issue in the repo myself https://github.com/Mariusthvdb/custom-ui/issues/105
